crates/adapters/src/controller/error.rs

```diff
@@ -35,38 +44,68 @@ pub enum ConfigError {
+    fn error_code(&self) -> Cow<'static, str> {
+        match self {
+            Self::ConfigParseError { .. } => Cow::from("ConfigParseError"),
+            Self::DuplicateInputEndpoint { .. } => Cow::from("DuplicateInputEndpoint"),
+            Self::DuplicateOutputEndpoint { .. } => Cow::from("DuplicateOutputEndpoint"),
+            Self::UnknownInputFormat { .. } => Cow::from("UnknownInputFormat"),
+            Self::UnknownOutputFormat { .. } => Cow::from("UnknownOutputFormat"),
+            Self::UnknownInputTransport { .. } => Cow::from("UnknownInputTransport"),
+            Self::UnknownOutputTransport { .. } => Cow::from("UnknownOutputTransport"),
+            Self::UnknownInputStream { .. } => Cow::from("UnknownInputStream"),
+            Self::UnknownOutputStream { .. } => Cow::from("UnknownOutputStream"),
+        }
+    }
+}
+
 impl Display for ConfigError {
     fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
         match self {
+            Self::ConfigParseError { error } => {
+                write!(f, "Failed to parse pipeline configuration: '{error}'")
+            }
             Self::DuplicateInputEndpoint { endpoint_name } => {
-                write!(f, "input endpoint '{endpoint_name}' already exists")
+                write!(f, "Input endpoint '{endpoint_name}' already exists")
             }
             Self::UnknownInputFormat { format_name } => {
-                write!(f, "unknown input format '{format_name}'")
+                write!(f, "Unknown input format '{format_name}'")
             }
             Self::UnknownInputTransport { transport_name } => {
-                write!(f, "unknown input transport '{transport_name}'")
+                write!(f, "Unknown input transport '{transport_name}'")
             }
             Self::DuplicateOutputEndpoint { endpoint_name } => {
-                write!(f, "output endpoint '{endpoint_name}' already exists")
+                write!(f, "Output endpoint '{endpoint_name}' already exists")
             }
             Self::UnknownOutputFormat { format_name } => {
-                write!(f, "unknown output format '{format_name}'")
+                write!(f, "Unknown output format '{format_name}'")
             }
             Self::UnknownOutputTransport { transport_name } => {
-                write!(f, "unknown output transport '{transport_name}'")
+                write!(f, "Unknown output transport '{transport_name}'")
             }
             Self::UnknownInputStream { stream_name } => {
-                write!(f, "unknown table '{stream_name}'")
+                write!(f, "Unknown table '{stream_name}'")
             }
             Self::UnknownOutputStream { stream_name } => {
-                write!(f, "unknown output table or view '{stream_name}'")
+                write!(f, "Unknown output table or view '{stream_name}'")
             }
         }
     }
 }
 
 impl ConfigError {
+    pub fn config_parse_error<E>(error: &E) -> Self
```

Comment:
can we include adapters in the pipeline manager, maybe this error type would be useful there

Reply:
I am going to implement strongly typed errors in pipeline manager next, but I think it should have its own config error type.